In [158]:
import pandas as pd
import numpy as np 
# read csv from github 'Sunqi'(enrollemnt)
enrollment = pd.read_csv('https://raw.githubusercontent.com/alexsong77/DAV-5400/master/Sunqi.csv')
enrollment

Unique IDof the institution Level of student  \
0                       100654    Undergraduate   
1                       100654         Graduate   
2                       100663    Undergraduate   
3                       100663         Graduate   
4                       100690    Undergraduate   
5                       100690         Graduate   
6                       100706    Undergraduate   
7                       100706         Graduate   

   American Indian or Alaska Native  Asian  Black or African American  \
0                                11     12                       4742   
1                                 4     11                        961   
2                                36    743                       3497   
3                                26    471                       1550   
4                                 0      2                        195   
5                                 2      1                        152   
6                                90    304                        813   
7                                23     71                        202   

   Hispanic or Latino  Native Hawaiian or Other Pacific Islander  White  \
0                  36                                         10    137   
1                   7                                          1    151   
2                 429                                          4   7515   
3                 289                                         11   6437   
4                  44                                          1     66   
5                  43                                          4     93   
6                 313                                          3   5055   
7                  64                                          0   1518   

   Two or more races  Unnamed: 9  Race/ethnicity unknown  Nonresident alien  
0                210         NaN                      11                 38  
1                 14         NaN                       3                 33  
2                522         NaN                     198                364  
3                269         NaN                     176                659  
4                  0         NaN                     123                  0  
5                  0         NaN                     142                  0  
6                129         NaN                     595                217  
7                 20         NaN                     115                365

**1.Purpose: Analyze the enrollment on different ethnicity by the level, or compare different institutions’ enrollment by ethnicity.**

In [159]:
# drop the none value 
enrollment.drop(columns=['Unnamed: 9'],inplace=True)
enrollment.dropna(how='all',inplace=True)

enrollment

Unique IDof the institution Level of student  \
0                       100654    Undergraduate   
1                       100654         Graduate   
2                       100663    Undergraduate   
3                       100663         Graduate   
4                       100690    Undergraduate   
5                       100690         Graduate   
6                       100706    Undergraduate   
7                       100706         Graduate   

   American Indian or Alaska Native  Asian  Black or African American  \
0                                11     12                       4742   
1                                 4     11                        961   
2                                36    743                       3497   
3                                26    471                       1550   
4                                 0      2                        195   
5                                 2      1                        152   
6                                90    304                        813   
7                                23     71                        202   

   Hispanic or Latino  Native Hawaiian or Other Pacific Islander  White  \
0                  36                                         10    137   
1                   7                                          1    151   
2                 429                                          4   7515   
3                 289                                         11   6437   
4                  44                                          1     66   
5                  43                                          4     93   
6                 313                                          3   5055   
7                  64                                          0   1518   

   Two or more races  Race/ethnicity unknown  Nonresident alien  
0                210                      11                 38  
1                 14                       3                 33  
2                522                     198                364  
3                269                     176                659  
4                  0                     123                  0  
5                  0                     142                  0  
6                129                     595                217  
7                 20                     115                365

In [160]:
# melt the wide format in to long format 
enrollment1 = enrollment.melt(id_vars=['Unique IDof the institution', 'Level of student'], value_vars=['American Indian or Alaska Native','Asian','Black or African American','Hispanic or Latino','Native Hawaiian or Other Pacific Islander','White','Two or more races','Race/ethnicity unknown','Nonresident alien'])
enrollment1.head()

Unique IDof the institution Level of student  \
0                       100654    Undergraduate   
1                       100654         Graduate   
2                       100663    Undergraduate   
3                       100663         Graduate   
4                       100690    Undergraduate   

                           variable  value  
0  American Indian or Alaska Native     11  
1  American Indian or Alaska Native      4  
2  American Indian or Alaska Native     36  
3  American Indian or Alaska Native     26  
4  American Indian or Alaska Native      0

In [161]:
#renamed the column name
enrollment1.columns = ['ID','Level of student','Ethnicity','Number of students']
enrollment1.head

<bound method NDFrame.head of         ID Level of student                         Ethnicity  \
0   100654    Undergraduate  American Indian or Alaska Native   
1   100654         Graduate  American Indian or Alaska Native   
2   100663    Undergraduate  American Indian or Alaska Native   
3   100663         Graduate  American Indian or Alaska Native   
4   100690    Undergraduate  American Indian or Alaska Native   
5   100690         Graduate  American Indian or Alaska Native   
6   100706    Undergraduate  American Indian or Alaska Native   
7   100706         Graduate  American Indian or Alaska Native   
8   100654    Undergraduate                             Asian   
9   100654         Graduate                             Asian   
10  100663    Undergraduate                             Asian   
11  100663         Graduate                             Asian   
12  100690    Undergraduate                             Asian   
13  100690         Graduate                             Asia

In [178]:
#group each dimensions by main categorizes 
enrollment2 = enrollment1.groupby([enrollment1['Level of student'],enrollment1['Ethnicity']],as_index=False)['Number of students'].sum()
enrollment2

Level of student                                  Ethnicity  \
0          Graduate           American Indian or Alaska Native   
1          Graduate                                      Asian   
2          Graduate                  Black or African American   
3          Graduate                         Hispanic or Latino   
4          Graduate  Native Hawaiian or Other Pacific Islander   
5          Graduate                          Nonresident alien   
6          Graduate                     Race/ethnicity unknown   
7          Graduate                          Two or more races   
8          Graduate                                      White   
9     Undergraduate           American Indian or Alaska Native   
10    Undergraduate                                      Asian   
11    Undergraduate                  Black or African American   
12    Undergraduate                         Hispanic or Latino   
13    Undergraduate  Native Hawaiian or Other Pacific Islander   
14    Undergraduate                          Nonresident alien   
15    Undergraduate                     Race/ethnicity unknown   
16    Undergraduate                          Two or more races   
17    Undergraduate                                      White   

    Number of students  
0                   55  
1                  554  
2                 2865  
3                  403  
4                   16  
5                 1057  
6                  436  
7                  303  
8                 8199  
9                  137  
10                1061  
11                9247  
12                 822  
13                  18  
14                 619  
15                 927  
16                 861  
17               12773

In [188]:
#group each dimensions by main categorizes "Level of student", "Ethnicity", and "Nummber of students"
enrollment2 = enrollment1.groupby([enrollment1['Level of student'],enrollment1['Ethnicity']])['Number of students'].sum().to_frame()
enrollment2

Number of students
Level of student Ethnicity                                                    
Graduate         American Indian or Alaska Native                           55
                 Asian                                                     554
                 Black or African American                                2865
                 Hispanic or Latino                                        403
                 Native Hawaiian or Other Pacific Islander                  16
                 Nonresident alien                                        1057
                 Race/ethnicity unknown                                    436
                 Two or more races                                         303
                 White                                                    8199
Undergraduate    American Indian or Alaska Native                          137
                 Asian                                                    1061
                 Black or African American                                9247
                 Hispanic or Latino                                        822
                 Native Hawaiian or Other Pacific Islander                  18
                 Nonresident alien                                         619
                 Race/ethnicity unknown                                    927
                 Two or more races                                         861
                 White                                                   12773

In [200]:
# find out which level of student and ethnicity has the highest number of enrollment students 
print(enrollment2.idxmax())
print(enrollment2['Number of students'].max())

Number of students    (Undergraduate, White)
dtype: object
12773


In [209]:
# find out which level of student and ethnicity has the lowest number of enrollment students 
print(enrollment2.idxmin())
print(enrollment2['Number of students'].min())

Number of students    (Graduate, Native Hawaiian or Other Pacific Is...
dtype: object
16


**Conclusion:for each level of student<br> "Undergraudete - White" has the highest number of enrollment students: 12773 <br>
                "Graduate - Native Hawaiian or Other Pacific Islander" has the lowest number of enrollment students:16**
            

In [203]:
#group each dimensions by main categorizes "ID", "Ethnicity", and "Nummber of students"
enrollment3 = enrollment1['Number of students'].groupby([enrollment1['ID'],enrollment1['Ethnicity']]).sum().to_frame()
enrollment3

Number of students
ID     Ethnicity                                                    
100654 American Indian or Alaska Native                           15
       Asian                                                      23
       Black or African American                                5703
       Hispanic or Latino                                         43
       Native Hawaiian or Other Pacific Islander                  11
       Nonresident alien                                          71
       Race/ethnicity unknown                                     14
       Two or more races                                         224
       White                                                     288
100663 American Indian or Alaska Native                           62
       Asian                                                    1214
       Black or African American                                5047
       Hispanic or Latino                                        718
       Native Hawaiian or Other Pacific Islander                  15
       Nonresident alien                                        1023
       Race/ethnicity unknown                                    374
       Two or more races                                         791
       White                                                   13952
100690 American Indian or Alaska Native                            2
       Asian                                                       3
       Black or African American                                 347
       Hispanic or Latino                                         87
       Native Hawaiian or Other Pacific Islander                   5
       Nonresident alien                                           0
       Race/ethnicity unknown                                    265
       Two or more races                                           0
       White                                                     159
100706 American Indian or Alaska Native                          113
       Asian                                                     375
       Black or African American                                1015
       Hispanic or Latino                                        377
       Native Hawaiian or Other Pacific Islander                   3
       Nonresident alien                                         582
       Race/ethnicity unknown                                    710
       Two or more races                                         149
       White                                                    6573

In [204]:
# find out which ID and ethnicity has the highest number of enrollment students 
print(enrollment3.idxmax())
print(enrollment3['Number of students'].max())

Number of students    (100663, White)
dtype: object
13952


In [208]:
# find out which ID and ethnicity has the lowest number of enrollment students 
print(enrollment3.idxmin())
print(enrollment3['Number of students'].min())

Number of students    (100690, Nonresident alien)
dtype: object
0


**Conclusion:for each ID of institution<br> "ID100663 - White" has the highest number of enrollment students: 13952 <br>
                "ID100690 - Nonresident alien" has the lowest number of enrollment students:0**

In [210]:
# read csv from github "Jeeho" (reviews)
reviews = pd.read_csv('https://raw.githubusercontent.com/alexsong77/DAV-5400/master/Jeeho%20-%20Destination%20Reviews.csv')
reviews

Unnamed: 0 User ID  Sports  Religious  Nature  Theatre  Shopping  Picnic
0           0  User 1       2         77      79       69        68      95
1           1  User 2       2         62      76       76        69      68
2           2  User 3       2         50      97       87        50      75
3           3  User 4       2         68      77       95        76      61
4           4  User 5       2         98      54       59        95      86

**2.Purpose:which class among the six has the highest frequency of reviews by the users**

In [46]:
# drop the none value 
reviews.drop(columns=['Unnamed: 0'],inplace=True)
reviews

User ID  Sports  Religious  Nature  Theatre  Shopping  Picnic
0  User 1       2         77      79       69        68      95
1  User 2       2         62      76       76        69      68
2  User 3       2         50      97       87        50      75
3  User 4       2         68      77       95        76      61
4  User 5       2         98      54       59        95      86

In [51]:
# melt the wide format in to long format
melted = pd.melt(reviews, ['User ID'])
melted.head()

User ID variable  value
0  User 1   Sports      2
1  User 2   Sports      2
2  User 3   Sports      2
3  User 4   Sports      2
4  User 5   Sports      2

In [66]:
#renamed the columns
melted.columns = ['User ID','Class','Value']
melted.head()

User ID   Class  Value
0  User 1  Sports      2
1  User 2  Sports      2
2  User 3  Sports      2
3  User 4  Sports      2
4  User 5  Sports      2

In [67]:
#groupby and sum each class 
melted.groupby('Class').sum()

Value
Class           
Nature       383
Picnic       385
Religious    355
Shopping     358
Sports        10
Theatre      386

**Conclusion: Class: "Theatre" has the highest frequency of reviews: 386**

In [211]:
# read csv from github “Natan” (grades)
grades = pd.read_csv('https://raw.githubusercontent.com/alexsong77/DAV-5400/master/Natan%20-%20Test%20Grades.csv')
grades

Name  Test 1  Test 2  Test 3  Test 4
0   Reggie Miller    97.0      80    87.0      91
1      Dan Bonner    76.0      68    80.0      81
2     Steve Smith    84.0      75     NaN      86
3  Charles Barkey     NaN      89     NaN      76
4     Kenny Smith    80.0      81    79.0      85
5    Chris Webber    75.0      75    76.0      92
6      Jalen Rose    80.0      85    84.0      83
7     Jim Jackson    92.0      91    94.0      95

**3.Purpose:<br> 1. Which student had the highest average test grades?<br>2. Which student had the greatest range of test grades?<br>3. Which test had the highest grades?**

 

In [26]:
# drop the none value 
grades.dropna(how='all',inplace=True)
grades

Name  Test 1  Test 2  Test 3  Test 4
0   Reggie Miller    97.0      80    87.0      91
1      Dan Bonner    76.0      68    80.0      81
2     Steve Smith    84.0      75     NaN      86
3  Charles Barkey     NaN      89     NaN      76
4     Kenny Smith    80.0      81    79.0      85
5    Chris Webber    75.0      75    76.0      92
6      Jalen Rose    80.0      85    84.0      83
7     Jim Jackson    92.0      91    94.0      95

In [68]:
# make up the non value by filling the previous columns
grades['Test 1'].fillna(method='ffill', inplace=True)
grades['Test 3'].fillna(method='ffill', inplace=True)
grades

Name  Test 1  Test 2  Test 3  Test 4
0   Reggie Miller    97.0      80    87.0      91
1      Dan Bonner    76.0      68    80.0      81
2     Steve Smith    84.0      75    80.0      86
3  Charles Barkey    84.0      89    80.0      76
4     Kenny Smith    80.0      81    79.0      85
5    Chris Webber    75.0      75    76.0      92
6      Jalen Rose    80.0      85    84.0      83
7     Jim Jackson    92.0      91    94.0      95

In [70]:
# melt the wide format in to long format 
melted1 = pd.melt(grades, ['Name'])
melted1.head()

Name variable  value
0   Reggie Miller   Test 1   97.0
1      Dan Bonner   Test 1   76.0
2     Steve Smith   Test 1   84.0
3  Charles Barkey   Test 1   84.0
4     Kenny Smith   Test 1   80.0

In [71]:
#renamed the columns
melted1.columns = ['Name','Test','Value']
melted1.head()

Name    Test  Value
0   Reggie Miller  Test 1   97.0
1      Dan Bonner  Test 1   76.0
2     Steve Smith  Test 1   84.0
3  Charles Barkey  Test 1   84.0
4     Kenny Smith  Test 1   80.0

In [87]:
#calculated the average test grades and groupby each categorizes by name, and then sorted the value 
name = melted1.groupby('Name').mean()
name.sort_values(by='Value',ascending=False)

Value
Name                 
Jim Jackson     93.00
Reggie Miller   88.75
Jalen Rose      83.00
Charles Barkey  82.25
Kenny Smith     81.25
Steve Smith     81.25
Chris Webber    79.50
Dan Bonner      76.25

**1.Jim Jackson had the highest average test grades**

In [92]:
# groupby and calculated each name ranges 
range1 = melted1.groupby('Name').max()['Value']-melted1.groupby('Name').min()['Value']
range1

Name
Charles Barkey    13.0
Chris Webber      17.0
Dan Bonner        13.0
Jalen Rose         5.0
Jim Jackson        4.0
Kenny Smith        6.0
Reggie Miller     17.0
Steve Smith       11.0
Name: Value, dtype: float64

In [213]:
# sorted the ranges of test grades
range1.sort_values(ascending=False)

Name
Reggie Miller     17.0
Chris Webber      17.0
Dan Bonner        13.0
Charles Barkey    13.0
Steve Smith       11.0
Kenny Smith        6.0
Jalen Rose         5.0
Jim Jackson        4.0
Name: Value, dtype: float64

**2.Reggie Miller and Chris Webber have the greatest range of test grades**

In [86]:
#Groupby and calculated the max of each test grades and then sorted by value 
test = melted1.groupby('Test').max()
test.sort_values(by='Value',ascending=False)

Name  Value
Test                      
Test 1  Steve Smith   97.0
Test 4  Steve Smith   95.0
Test 3  Steve Smith   94.0
Test 2  Steve Smith   91.0

**3.Test4 had the highest grades**